### PR TITLE
Using Backward-cpp for stacktraces under Linux

### PR DESCRIPTION
**Pull request by ElNounch**
With [Backward-cpp](https://github.com/bombela/backward-cpp) :
Turning

```
[c6fc1e936562b741|22:29:09] Assertion failed: !"Unhandled block face!", file /home/elnounch/cuberite/src/Blocks/BlockTrapdoor.h, line 82
/home/cuberite/Server/Cuberite_debug(_Z15PrintStackTracev+0x32)[0xb532b2]
/home/cuberite/Server/Cuberite_debug(_ZN21cBlockTrapdoorHandler19BlockFaceToMetaDataE10eBlockFace+0x67)[0xca9a65]
/home/cuberite/Server/Cuberite_debug(_ZN21cBlockTrapdoorHandler25GetPlacementBlockTypeMetaER15cChunkInterfaceP7cPlayeriii10eBlockFaceiiiRhS5_+0x35)[0xca99d7]
/home/cuberite/Server/Cuberite_debug(_ZN12cItemHandler25GetPlacementBlockTypeMetaEP6cWorldP7cPlayeriii10eBlockFaceiiiRhS5_+0x139)[0xc80289]
/home/cuberite/Server/Cuberite_debug(_ZN12cItemHandler16GetBlocksToPlaceER6cWorldR7cPlayerRK5cItemiii10eBlockFaceiiiRSt6vectorI9sSetBlockSaIS9_EE+0x85)[0xc7f959]
/home/cuberite/Server/Cuberite_debug(_ZN12cItemHandler13OnPlayerPlaceER6cWorldR7cPlayerRK5cItemiii10eBlockFaceiii+0x38f)[0xc7f4a1]
/home/cuberite/Server/Cuberite_debug(_ZN13cClientHandle16HandleRightClickEiii10eBlockFaceiiiRK5cItem+0xb92)[0xa6a2f6]
/home/cuberite/Server/Cuberite_debug(_ZN12cProtocol19022HandlePacketBlockPlaceER11cByteBuffer+0x14e)[0xce9afc]
/home/cuberite/Server/Cuberite_debug(_ZN12cProtocol19012HandlePacketER11cByteBufferj+0x3fd)[0xce87ff]
/home/cuberite/Server/Cuberite_debug(_ZN12cProtocol19015AddReceivedDataEPKcm+0x983)[0xce7d91]
/home/cuberite/Server/Cuberite_debug(_ZN12cProtocol19012DataReceivedEPKcm+0x101)[0xcde135]
/home/cuberite/Server/Cuberite_debug(_ZN19cProtocolRecognizer12DataReceivedEPKcm+0x316)[0xcf55c2]
/home/cuberite/Server/Cuberite_debug(_ZN13cClientHandle4TickEf+0xfb)[0xa6bf0b]
/home/cuberite/Server/Cuberite_debug(_ZN6cWorld11TickClientsEf+0x3fe)[0xb1917a]
/home/cuberite/Server/Cuberite_debug(_ZN6cWorld4TickENSt6chrono8durationIlSt5ratioILl1ELl1000EEEES4_+0x40e)[0xb182c4]
/home/cuberite/Server/Cuberite_debug(_ZN6cWorld11cTickThread7ExecuteEv+0xb7)[0xb0e1b7]
/home/cuberite/Server/Cuberite_debug(_ZN9cIsThread9DoExecuteEv+0x2f)[0xb47cd7]
/home/cuberite/Server/Cuberite_debug(_ZNKSt12_Mem_fn_baseIM9cIsThreadFvvELb1EEclIIEvEEvPS0_DpOT_+0x65)[0xb4946b]
/home/cuberite/Server/Cuberite_debug(_ZNSt12_Bind_simpleIFSt7_Mem_fnIM9cIsThreadFvvEEPS1_EE9_M_invokeIILm0EEEEvSt12_Index_tupleIIXspT_EEE+0x43)[0xb493ff]
/home/cuberite/Server/Cuberite_debug(_ZNSt12_Bind_simpleIFSt7_Mem_fnIM9cIsThreadFvvEEPS1_EEclEv+0x2c)[0xb49306]
/home/cuberite/Server/Cuberite_debug(_ZNSt6thread5_ImplISt12_Bind_simpleIFSt7_Mem_fnIM9cIsThreadFvvEEPS3_EEE6_M_runEv+0x1c)[0xb49296]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xb8c80)[0x7f415b695c80]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x76fa)[0x7f415c7fa6fa]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d)[0x7f415adfbb5d]
Cuberite_debug: /home/elnounch/cuberite/src/Blocks/BlockTrapdoor.h:82: static NIBBLETYPE cBlockTrapdoorHandler::BlockFaceToMetaData(eBlockFace): Assertion `0' failed.
Aborted
```

into

```
[6e33c990bf571326|08:41:06] Assertion failed: !"Unhandled block face!", file /home/elnounch/cuberite/src/Blocks/BlockTrapdoor.h, line 82
Stack trace (most recent call last) in thread 20350:
#26   Object "[0xffffffffffffffff]", at 0xffffffffffffffff, in 
#25   Object "/lib/x86_64-linux-gnu/libc.so.6", at 0x7fc542451b5c, in clone
#24   Object "/lib/x86_64-linux-gnu/libpthread.so.0", at 0x7fc543e506f9, in 
#23   Object "/usr/lib/x86_64-linux-gnu/libstdc++.so.6", at 0x7fc542cebc7f, in 
#22   Object "/home/cuberite/Server/Cuberite_debug", at 0xb5bd1b, in std::thread::_Impl<std::_Bind_simple<std::_Mem_fn<void (cIsThread::*)()> (cIsThread*)> >::_M_run()
#21   Object "/home/cuberite/Server/Cuberite_debug", at 0xb5bd8b, in std::_Bind_simple<std::_Mem_fn<void (cIsThread::*)()> (cIsThread*)>::operator()()
#20   Object "/home/cuberite/Server/Cuberite_debug", at 0xb5be84, in void std::_Bind_simple<std::_Mem_fn<void (cIsThread::*)()> (cIsThread*)>::_M_invoke<0ul>(std::_Index_tuple<0ul>)
#19   Object "/home/cuberite/Server/Cuberite_debug", at 0xb5bef0, in void std::_Mem_fn_base<void (cIsThread::*)(), true>::operator()<, void>(cIsThread*) const
#18   Object "/home/cuberite/Server/Cuberite_debug", at 0xb5a75c, in cIsThread::DoExecute()
#17   Object "/home/cuberite/Server/Cuberite_debug", at 0xb20d3c, in cWorld::cTickThread::Execute()
#16   Object "/home/cuberite/Server/Cuberite_debug", at 0xb2ae49, in cWorld::Tick(std::chrono::duration<long, std::ratio<1l, 1000l> >, std::chrono::duration<long, std::ratio<1l, 1000l> >)
#15   Object "/home/cuberite/Server/Cuberite_debug", at 0xb2bcff, in cWorld::TickClients(float)
#14   Object "/home/cuberite/Server/Cuberite_debug", at 0xa7ea90, in cClientHandle::Tick(float)
#13   Object "/home/cuberite/Server/Cuberite_debug", at 0xd0efad, in cProtocolRecognizer::DataReceived(char const*, unsigned long)
#12   Object "/home/cuberite/Server/Cuberite_debug", at 0xcf7b20, in cProtocol190::DataReceived(char const*, unsigned long)
#11   Object "/home/cuberite/Server/Cuberite_debug", at 0xd0177c, in cProtocol190::AddReceivedData(char const*, unsigned long)
#10   Object "/home/cuberite/Server/Cuberite_debug", at 0xd021ea, in cProtocol190::HandlePacket(cByteBuffer&, unsigned int)
#9    Object "/home/cuberite/Server/Cuberite_debug", at 0xd034e7, in cProtocol190::HandlePacketBlockPlace(cByteBuffer&)
#8    Object "/home/cuberite/Server/Cuberite_debug", at 0xa7ce7c, in cClientHandle::HandleRightClick(int, int, int, eBlockFace, int, int, int, cItem const&)
#7    Object "/home/cuberite/Server/Cuberite_debug", at 0xc991b0, in cItemHandler::OnPlayerPlace(cWorld&, cPlayer&, cItem const&, int, int, int, eBlockFace, int, int, int)
#6    Object "/home/cuberite/Server/Cuberite_debug", at 0xc99668, in cItemHandler::GetBlocksToPlace(cWorld&, cPlayer&, cItem const&, int, int, int, eBlockFace, int, int, int, std::vector<sSetBlock, std::allocator<sSetBlock> >&)
#5    Object "/home/cuberite/Server/Cuberite_debug", at 0xc99f98, in cItemHandler::GetPlacementBlockTypeMeta(cWorld*, cPlayer*, int, int, int, eBlockFace, int, int, int, unsigned char&, unsigned char&)
#4    Object "/home/cuberite/Server/Cuberite_debug", at 0xcc34d0, in cBlockTrapdoorHandler::GetPlacementBlockTypeMeta(cChunkInterface&, cPlayer*, int, int, int, eBlockFace, int, int, int, unsigned char&, unsigned char&)
#3    Object "/home/cuberite/Server/Cuberite_debug", at 0xcc355e, in cBlockTrapdoorHandler::BlockFaceToMetaData(eBlockFace)
#2    Object "/home/cuberite/Server/Cuberite_debug", at 0xb65d43, in PrintStackTrace()
#1    Object "/home/cuberite/Server/Cuberite_debug", at 0xb6633b, in backward::StackTraceImpl<backward::system_tag::linux_tag>::load_here(unsigned long)
#0    Object "/home/cuberite/Server/Cuberite_debug", at 0xb67d59, in unsigned long backward::details::unwind<backward::StackTraceImpl<backward::system_tag::linux_tag>::callback>(backward::StackTraceImpl<backward::system_tag::linux_tag>::callback, unsigned long)
Cuberite_debug: /home/elnounch/cuberite/src/Blocks/BlockTrapdoor.h:82: static NIBBLETYPE cBlockTrapdoorHandler::BlockFaceToMetaData(eBlockFace): Assertion `0' failed.
Aborted (core dumped)
```

or even (debug and bfd library) :

```
[6575dbe2d41e5a55|22:20:54] Assertion failed: !"Unhandled block face!", file /home/elnounch/cuberite/src/Blocks/BlockTrapdoor.h, line 82
Stack trace (most recent call last) in thread 27521:
#26   Object "", at 0xffffffffffffffff, in 
#25   Source "/build/glibc-GKVZIf/glibc-2.23/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S", line 109, in  [0x7f1b6b290b5c]
#24   Object "/lib/x86_64-linux-gnu/libpthread.so.0", at 0x7f1b6cc8f6f9, in start_thread
#23   Object "/usr/lib/x86_64-linux-gnu/libstdc++.so.6", at 0x7f1b6bb2ac7f, in std::this_thread::__sleep_for(std::chrono::duration<long, std::ratio<1l, 1l> >, std::chrono::duration<long, std::ratio<1l, 1000000000l> >)
#22   Source "/usr/include/c++/5/thread", line 115, in std::thread::_Impl<std::_Bind_simple<std::_Mem_fn<void (cIsThread::*)()> (cIsThread*)> >::_M_run() [0xb67465]
        112:    { }
        113: 
        114:    void
      > 115:    _M_run() { _M_func(); }
        116:       };
        117: 
        118:   private:
#21   Source "/usr/include/c++/5/functional", line 1520, in std::_Bind_simple<std::_Mem_fn<void (cIsThread::*)()> (cIsThread*)>::operator()() [0xb674d5]
       1517:       operator()()
       1518:       {
       1519:         typedef typename _Build_index_tuple<sizeof...(_Args)>::__type _Indices;
      >1520:         return _M_invoke(_Indices());
       1521:       }
       1522: 
       1523:     private:
#20   Source "/usr/include/c++/5/functional", line 1531, in void std::_Bind_simple<std::_Mem_fn<void (cIsThread::*)()> (cIsThread*)>::_M_invoke<0ul>(std::_Index_tuple<0ul>) [0xb675ce]
       1528:      // std::bind always forwards bound arguments as lvalues,
       1529:      // but this type can call functions which only accept rvalues.
       1530:           return std::forward<_Callable>(std::get<0>(_M_bound))(
      >1531:               std::forward<_Args>(std::get<_Indices+1>(_M_bound))...);
       1532:         }
       1533: 
       1534:       std::tuple<_Callable, _Args...> _M_bound;
#19   Source "/usr/include/c++/5/functional", line 600, in void std::_Mem_fn_base<void (cIsThread::*)(), true>::operator()<, void>(cIsThread*) const [0xb6763a]
        597:                           _CheckArgs<_Pack<_Args...>>>>
        598:    result_type
        599:    operator()(_Class* __object, _Args&&... __args) const
      > 600:    { return (__object->*_M_pmf)(std::forward<_Args>(__args)...); }
        601: 
        602:       // Handle smart pointers, references and pointers to derived
        603:       template<typename _Tp, typename... _Args, typename _Req
#18   Source "/home/elnounch/cuberite/src/OSSupport/IsThread.cpp", line 74, in cIsThread::DoExecute() [0xb65ea6]
         71: void cIsThread::DoExecute(void)
         72: {
         73:    m_evtStart.Wait();
      >  74:    Execute();
         75: }
#17   Source "/home/elnounch/cuberite/src/World.cpp", line 111, in cWorld::cTickThread::Execute() [0xb2c386]
        108:    {
        109:        auto NowTime = std::chrono::steady_clock::now();
        110:        auto WaitTime = std::chrono::duration_cast<std::chrono::milliseconds>(NowTime - LastTime);
      > 111:        m_World.Tick(WaitTime, TickTime);
        112:        TickTime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - NowTime);
        113: 
        114:        if (TickTime < cTickTime(1))
#16   Source "/home/elnounch/cuberite/src/World.cpp", line 1052, in cWorld::Tick(std::chrono::duration<long, std::ratio<1l, 1000l> >, std::chrono::duration<long, std::ratio<1l, 1000l> >) [0xb36493]
       1049:    TickMobs(a_Dt);
       1050:    m_MapManager.TickMaps();
       1051: 
      >1052:    TickClients(static_cast<float>(a_Dt.count()));
       1053:    TickQueuedBlocks();
       1054:    TickQueuedTasks();
#15   Source "/home/elnounch/cuberite/src/World.cpp", line 1283, in cWorld::TickClients(float) [0xb37349]
       1280:                itr = m_Clients.erase(itr);
       1281:                continue;
       1282:            }
      >1283:            (*itr)->Tick(a_Dt);
       1284:            ++itr;
       1285:        }  // for itr - m_Clients[]
       1286:    }
#14   Source "/home/elnounch/cuberite/src/ClientHandle.cpp", line 1958, in cClientHandle::Tick(float) [0xa8a0da]
       1955:    }
       1956:    if (!IncomingData.empty())
       1957:    {
      >1958:        m_Protocol->DataReceived(IncomingData.data(), IncomingData.size());
       1959:    }
       1960: 
       1961:    // Send any queued outgoing data:
#13   Source "/home/elnounch/cuberite/src/Protocol/ProtocolRecognizer.cpp", line 114, in cProtocolRecognizer::DataReceived(char const*, unsigned long) [0xd1dba9]
        111:    }
        112:    else
        113:    {
      > 114:        m_Protocol->DataReceived(a_Data, a_Size);
        115:    }
        116: }
#12   Source "/home/elnounch/cuberite/src/Protocol/Protocol19x.cpp", line 176, in cProtocol190::DataReceived(char const*, unsigned long) [0xd0671c]
        173:    }
        174:    else
        175:    {
      > 176:        AddReceivedData(a_Data, a_Size);
        177:    }
        178: }
#11   Source "/home/elnounch/cuberite/src/Protocol/Protocol19x.cpp", line 1956, in cProtocol190::AddReceivedData(char const*, unsigned long) [0xd10378]
       1953:            );
       1954:        }
       1955: 
      >1956:        if (!HandlePacket(bb, PacketType))
       1957:        {
       1958:            // Unknown packet, already been reported, but without the length. Log the length here:
       1959:            LOGWARNING("Protocol 1.9: Unhandled packet: type 0x%x, state %d, length %u", PacketType, m_State, PacketLen);
#10   Source "/home/elnounch/cuberite/src/Protocol/Protocol19x.cpp", line 2083, in cProtocol190::HandlePacket(cByteBuffer&, unsigned int) [0xd10de6]
       2080:                case 0x19: HandlePacketUpdateSign             (a_ByteBuffer); return true;
       2081:                case 0x1a: HandlePacketAnimation              (a_ByteBuffer); return true;
       2082:                case 0x1b: break;  // Spectate?
      >2083:                case 0x1c: HandlePacketBlockPlace             (a_ByteBuffer); return true;
       2084:                case 0x1d: HandlePacketUseItem                (a_ByteBuffer); return true;
       2085:            }
       2086:            break;
#9    Source "/home/elnounch/cuberite/src/Protocol/Protocol19x.cpp", line 2315, in cProtocol190::HandlePacketBlockPlace(cByteBuffer&) [0xd120e3]
       2312:    HANDLE_READ(a_ByteBuffer, ReadBEUInt8, UInt8, CursorX);
       2313:    HANDLE_READ(a_ByteBuffer, ReadBEUInt8, UInt8, CursorY);
       2314:    HANDLE_READ(a_ByteBuffer, ReadBEUInt8, UInt8, CursorZ);
      >2315:    m_Client->HandleRightClick(BlockX, BlockY, BlockZ, FaceIntToBlockFace(Face), CursorX, CursorY, CursorZ, m_Client->GetPlayer()->GetEquippedItem());
       2316: }
#8    Source "/home/elnounch/cuberite/src/ClientHandle.cpp", line 1445, in cClientHandle::HandleRightClick(int, int, int, eBlockFace, int, int, int, cItem const&) [0xa884c5]
       1443:    if (ItemHandler->IsPlaceable() && (a_BlockFace != BLOCK_FACE_NONE))
       1444:    {
      >1445:        if (!ItemHandler->OnPlayerPlace(*World, *m_Player, Equipped, a_BlockX, a_BlockY, a_BlockZ, a_BlockFace, a_CursorX, a_CursorY, a_CursorZ))
       1446:        {
       1447:            // Placement failed, bail out
       1448:            return;
#7    Source "/home/elnounch/cuberite/src/Items/ItemHandler.cpp", line 381, in cItemHandler::OnPlayerPlace(cWorld&, cPlayer&, cItem const&, int, int, int, eBlockFace, int, int, int) [0xca7a88]
        379:    // Get all the blocks to place:
        380:    sSetBlockVector blocks;
      > 381:    if (!GetBlocksToPlace(a_World, a_Player, a_EquippedItem, a_BlockX, a_BlockY, a_BlockZ, a_BlockFace, a_CursorX, a_CursorY, a_CursorZ, blocks))
        382:    {
        383:        // Handler refused the placement, send that information back to the client:
        384:        for (const auto & blk: blocks)
#6    Source "/home/elnounch/cuberite/src/Items/ItemHandler.cpp", line 445, in cItemHandler::GetBlocksToPlace(cWorld&, cPlayer&, cItem const&, int, int, int, eBlockFace, int, int, int, std::vector<sSetBlock, std::allocator<sSetBlock> >&) [0xca7f40]
        442: {
        443:    BLOCKTYPE BlockType;
        444:    NIBBLETYPE BlockMeta;
      > 445:    if (!GetPlacementBlockTypeMeta(&a_World, &a_Player, a_BlockX, a_BlockY, a_BlockZ, a_BlockFace, a_CursorX, a_CursorY, a_CursorZ, BlockType, BlockMeta))
        446:    {
        447:        return false;
        448:    }
#5    Source "/home/elnounch/cuberite/src/Items/ItemHandler.cpp", line 817, in cItemHandler::GetPlacementBlockTypeMeta(cWorld*, cPlayer*, int, int, int, eBlockFace, int, int, int, unsigned char&, unsigned char&) [0xca8870]
        814:        a_BlockX, a_BlockY, a_BlockZ, a_BlockFace,
        815:        a_CursorX, a_CursorY, a_CursorZ,
        816:        a_BlockType, a_BlockMeta
      > 817:    );
        818: }
#4    Source "/home/elnounch/cuberite/src/Blocks/BlockTrapdoor.h", line 61, in cBlockTrapdoorHandler::GetPlacementBlockTypeMeta(cChunkInterface&, cPlayer*, int, int, int, eBlockFace, int, int, int, unsigned char&, unsigned char&) [0xcd1fbe]
         58:    ) override
         59:    {
         60:        a_BlockType = m_BlockType;
      >  61:        a_BlockMeta = BlockFaceToMetaData(a_BlockFace);
         62: 
         63:        if (a_CursorY > 7)
         64:        {
#3    Source "/home/elnounch/cuberite/src/Blocks/BlockTrapdoor.h", line 82, in cBlockTrapdoorHandler::BlockFaceToMetaData(eBlockFace) [0xcd204c]
         79:            case BLOCK_FACE_YM:
         80:            case BLOCK_FACE_YP:
         81:            {
      >  82:                ASSERT(!"Unhandled block face!");
         83:                return 0;
         84:            }
         85:        }
#2    Source "/home/elnounch/cuberite/src/OSSupport/StackTrace.cpp", line 42, in PrintStackTrace() [0xb7148d]
         39:    #else
         40:        #ifdef __linux
         41:            backward::StackTrace st; 
      >  42:            st.load_here(32);
         43:            backward::Printer p;
         44:            p.color = true;
         45:            p.address = true;
#1    Source "/home/elnounch/cuberite/src/OSSupport/backward.hpp", line 615, in backward::StackTraceImpl<backward::system_tag::linux_tag>::load_here(unsigned long) [0xb71a45]
        612:            return 0;
        613:        }
        614:        _stacktrace.resize(depth);
      > 615:        size_t trace_cnt = details::unwind(callback(*this), depth);
        616:        _stacktrace.resize(trace_cnt);
        617:        skip_n_firsts(0);
        618:        return size();
#0    Source "/home/elnounch/cuberite/src/OSSupport/backward.hpp", line 599, in unsigned long backward::details::unwind<backward::StackTraceImpl<backward::system_tag::linux_tag>::callback>(backward::StackTraceImpl<backward::system_tag::linux_tag>::callback, unsigned long) [0xb7422f]
        596: template <typename F>
        597: size_t unwind(F f, size_t depth) {
        598:    Unwinder<F> unwinder;
      > 599:    return unwinder(f, depth);
        600: }
        601: 
        602: } // namespace details
Cuberite_debug: /home/elnounch/cuberite/src/Blocks/BlockTrapdoor.h:82: static NIBBLETYPE cBlockTrapdoorHandler::BlockFaceToMetaData(eBlockFace): Assertion `0' failed.
Aborted
```

### CODE REVIEW

**Comment by ElNounch** [comment]
Let's consider each commit separetely :sweat_smile:

**Comment by madmaxoft** [comment]
The library should have been added as a submodule. This means creating a separate repo for it (if it doesn't have one on GitHub yet) and referencing that repo instead of adding the sources.

**Comment by madmaxoft** [comment]
Additionally, you won't need to modify `CheckBasicStyle.lua` when the external code is in a submodule.

File: src/OSSupport/StackTrace.cpp [hunk 1]
```diff
@@ -20,25 +20,34 @@
...
+	p.color = true;
+	p.address = true;
+	p.print(st, stderr);
+#elif defined(__GLIBC__)
+	// Use the backtrace() function to get and output the stackTrace:
+	// Code adapted from https://stackoverflow.com/questions/77005/how-to-generate-a-stacktrace-when-my-gcc-c-app-crashes
+	void * stackTrace[30];
+	auto numItems = backtrace(stackTrace, ARRAYCOUNT(stackTrace));
+	backtrace_symbols_fd(stackTrace, numItems, STDERR_FILENO);
+#endif
```
Comment (madmaxoft) [inline src/OSSupport/StackTrace.cpp]: The preprocessor directives should be indented properly (as they used to be).

**Comment by madmaxoft** [comment]
Since the backward-cpp library seems to be failing our CI in its cmake, you might want to fork the repo to the Cuberite organization, and make the necessary changes there to make it compile.

**Comment by ElNounch** [comment]
Saw sub-repo as an [interesting alternative](https://codingkilledthecat.wordpress.com/2012/04/28/why-your-company-shouldnt-use-git-submodules/) to submodules. ^^
Plus the CI's failure might be due to obsolete CMake on Travis. :wink:
Taking an easier path -> https://github.com/cuberite/cuberite/pull/3347

**Pull request closed without merging**